src/network/bpf/sysctl_monitor/sysctl-monitor.bpf.c

```diff
@@ -0,0 +1,92 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "vmlinux.h"
+
+#include <bpf/bpf_helpers.h>
+
+#include "sysctl-write-event.h"
+
+struct {
+        __uint(type, BPF_MAP_TYPE_RINGBUF);
+        __uint(max_entries, 256 * 1024);
+} written_sysctls SEC(".maps");
+
+static bool my_streq(const char *s1, const char *s2, int l)
+{
+        for (int i = 0; i < l; i++) {
+                if (s1[i] != s2[i])
+                        return false;
+                if (!s1[i])
+                        return true;
+        }
+        return true;
+}
+
+struct str {
+        char *s;
+        int l;
+};
+
+static long cut_last(int i, struct str *str)
+{
+        char *s;
+
+        i = str->l - i - 1;
+        s = str->s + i;
+
+        // Sanity checks for the preverifier
+        if (i < 0 || i >= str->l)
+                return 1;
+
+        if (*s == 0)
+                return 0;
+
+        if (*s == '\n' || *s == '\r' || *s == ' ' || *s == '\t') {
+                *s = 0;
+
+                return 0;
+        }
+
+        return 1;
+}
+
+// Cut off trailing whitespace and newlines
+static void chop(char *s, int l)
+{
+        struct str str = { s, l };
+
+        bpf_loop(l, cut_last, &str, 0);
+}
+
+SEC("cgroup/sysctl")
+int sysctl_monitor(struct bpf_sysctl *ctx)
+{
+        struct sysctl_write_event we = { };
+
+        // Allow reads
+        if (!ctx->write)
+                return 1;
+
+        // Only monitor net/
+        bpf_sysctl_get_name(ctx, we.name, sizeof(we.name), 0);
+        if (bpf_strncmp(we.name, 4, "net/"))
+                return 1;
+
+        we.pid = bpf_get_current_pid_tgid() >> 32;
+        bpf_get_current_comm(we.comm, sizeof(we.comm));
+        bpf_sysctl_get_current_value(ctx, we.current, sizeof(we.current));
+        bpf_sysctl_get_new_value(ctx, we.newvalue, sizeof(we.newvalue));
+
+        // Both the kernel and userspace applications add a newline at the end,
+        // remove it from both strings
+        chop(we.current, sizeof(we.current));
+        chop(we.newvalue, sizeof(we.newvalue));
+
+        // If new value differs, send the event
+        if (!my_streq(we.current, we.newvalue, sizeof(we.current)))
```

Comment:
why not use bpf_strncmp() here too?

Reply:
`bpf_strncmp()` requires the second string to be const

Reply:
bpf_strncmp() fails like this with a non read-only value:

```
[   39.648359] systemd-networkd[1082]: libbpf: prog 'sysctl_monitor': BPF program load failed: Permission denied
[   39.648374] systemd-networkd[1082]: libbpf: prog 'sysctl_monitor': -- BEGIN PROG LOAD LOG --
[   39.648380] systemd-networkd[1082]: 0: R1=ctx() R10=fp0
[   39.648386] systemd-networkd[1082]: ; int sysctl_monitor(struct bpf_sysctl *ctx) { @ sysctl-monitor.bpf.c:59
[   39.648393] systemd-networkd[1082]: 0: (bf) r6 = r1                       ; R1=ctx() R6_w=ctx()
[   39.648399] systemd-networkd[1082]: 1: (b7) r1 = 0                        ; R1_w=0
[   39.648404] systemd-networkd[1082]: ; struct sysctl_write_event we = { }; @ sysctl-monitor.bpf.c:60
[   39.648412] systemd-networkd[1082]: 2: (63) *(u32 *)(r10 -24) = r1        ; R1_w=0 R10=fp0 fp-24=????0
[...]
[   39.652253] systemd-networkd[1082]: libbpf: prog 'sysctl_monitor': failed to load: -13
[   39.652961] systemd-networkd[1082]: libbpf: failed to load object 'sysctl_monitor_bpf'
[   39.652975] systemd-networkd[1082]: libbpf: failed to load BPF skeleton 'sysctl_monitor_bpf': -13
[   39.654205] systemd-networkd[1082]: Failed to load sysctl monitor BPF program: Permission denied
```